katdal/categorical.py

```diff
@@ -591,69 +654,92 @@ def sensor_to_categorical(sensor_timestamps, sensor_values, dump_midtimes, dump_
-    # This also gets rid of excess events before the first dump and after the last dump
-    non_empty = np.nonzero(np.diff(events) > 0)[0]
-    sensor_values, events = sensor_values[non_empty], np.r_[events[non_empty], len(dump_midtimes)]
-    # Force first dump to have valid sensor value (use initial value or first proper value advanced to start)
-    if events[0] != 0 and initial_value is not None:
-        if wrapped_values:
-            initial_value = ComparableArrayWrapper(initial_value)
-        sensor_values, events = np.r_[[initial_value], sensor_values], np.r_[0, events]
-    events[0] = 0
+    # Convert sensor event times to dump indices by picking the dump during
+    # which each sensor event occurred (more precisely: closest dump centroid).
+    # Last event is fixed at one-past-last-dump to indicate end of last segment.
+    events = np.r_[dump_endtimes.searchsorted(sensor_timestamps) - 1, num_dumps]
+    # Get rid of excess events before the first dump and after the last dump
+    # The dump index of prior events is -1 and of later events is `num_dumps`
+    first_proper_event = events.searchsorted(-1, side='right')
+    # Shift the final prior event (if any) to the start of the first dump
+    if first_proper_event > 0:
+        first_proper_event -= 1
+        events[first_proper_event] = 0
+    one_past_last_event = events.searchsorted(num_dumps)
+    within_dumps = slice(first_proper_event, one_past_last_event)
+    sensor_values = sensor_values[within_dumps]
+    events = np.r_[events[within_dumps], num_dumps]
     # Apply optional transform to sensor values
     if transform is not None:
         if wrapped_values:
-            unwrapped_transform = transform
-            transform = lambda value: ComparableArrayWrapper(unwrapped_transform(value.unwrapped))  # noqa: E731
+            orig_transform = transform
+            def transform(value):   # noqa: E301
+                """Unwrap wrapped value, transform and rewrap."""
+                return ComparableArrayWrapper(orig_transform(value.unwrapped))
         sensor_values = np.array([transform(y) for y in sensor_values])
-    # Discard sensor events that do not change the (transformed) sensor value (i.e. that repeat the previous value)
+    # Force first dump to have valid sensor value
+    # (insert initial value or let the first proper value apply from the start)
+    if events[0] != 0 and initial_value is not None:
+        if wrapped_values:
+            initial_value = ComparableArrayWrapper(initial_value)
+        sensor_values = np.r_[[initial_value], sensor_values]
+        events = np.r_[0, events]
+    events[0] = 0
+    # Clean up dump->event mapping, taking into account greedy values
+    greedy_values = () if greedy_values is None else greedy_values
+    greedy = [value in greedy_values for value in sensor_values]
+    cleaned_up = [event for event in _single_event_per_dump(events, greedy)]
```

Comment:
Could also write a `list(_single_event_per_dump(events, greedy))`. May just be a matter of taste.